StripeCore/StripeCore/Source/Attestation/StripeAttestBackend.swift

```diff
@@ -6,23 +6,37 @@
 import Foundation
 
 @_spi(STP) public protocol StripeAttestBackend {
-    func getChallenge(appId: String, deviceId: String, keyId: String) async throws -> String
+    func getChallenge(appId: String, deviceId: String, keyId: String) async throws -> StripeChallengeResponse
     func attest(appId: String, deviceId: String, keyId: String, attestation: Data) async throws
 }
 
+@_spi(STP) public struct StripeChallengeResponse: Decodable {
+    let challenge: String
+    let initial_attestation_required: Bool
+}
+
 @_spi(STP) public class StripeAPIAttestationBackend: StripeAttestBackend {
     let apiClient: STPAPIClient
 
     @_spi(STP) public init(apiClient: STPAPIClient) {
         self.apiClient = apiClient
     }
 
-    public func attest(appId: String, deviceId: String, keyId: String, attestation: Data) async throws {
-        stpAssertionFailure("Not implemented")
+    @_spi(STP) public func attest(appId: String, deviceId: String, keyId: String, attestation: Data) async throws {
+        let _: EmptyResponse = try await withCheckedThrowingContinuation { continuation in
+            apiClient.post(resource: "mobile_sdk_attestation/ios_attest", parameters: ["app_id": appId, "device_id": deviceId, "key_id": keyId, "attestation_object": attestation.base64EncodedString()]) { result in
```

Comment:
Couple suggestions here that I think would help clean this up a tiny bit; 

1. Could the base parameters (`appId`, `deviceId`, `keyId`) be set on class / protocol? This way they don't need to be passed in both methods requesting them. Though this wouldn't really work if the parameters could change between the initialization time and the function call time. Something like:

```swift
protocol StripeAttestBackend {
    var appId: String { get }
    var deviceId: String { get }
    var keyId: String { get }

    func getChallenge() async throws -> StripeChallengeResponse
    func attest(attestation: Data) async throws
}
```

2. It'd be nice to maybe encapsulate some of this endpoint logic into an enum along the lines of: 

```swift
enum Endpoint {
    case attest(attestation: Data)
    case challenge
    
    var path: String {
        switch self {
        case .attest: "mobile_sdk_attestation/ios_attest"
        case .challenge: "mobile_sdk_attestation/ios_challenge"
        }
    }
}
```

3. That alone would be nice, but it would pair well with a function on `StripeAPIAttestationBackend` to build the parameters object:

```swift
private func parameters(for endpoint: Endpoint) -> [String: String] {
    var baseParams: [String: String] = [
        "app_id": appId,
        "device_id": deviceId,
        "key_id": keyId,
    ]
    
    switch endpoint {
    case .attest(let attestation):
        let attestationObject = ["attestation_object": attestation.base64EncodedString()]
        return baseParams.merging(attestationObject) { (current, _) in current }
    case .challenge:
        return baseParams
    }
}
```

4. And finally, these two functions would be implemented as:

```swift
public func attest(attestation: Data) async throws {
    let _: EmptyResponse = try await withCheckedThrowingContinuation { continuation in
        let endpoint = Endpoint.attest(attestation: attestation)
        apiClient.post(resource: endpoint.path, parameters: parameters(for: endpoint))) { result in
            continuation.resume(with: result)
        }
    }
}
```

Feel free to cherry pick anything you like or even disregard! It really all comes down to personal preference

Reply:
Thanks! I'm thinking about this, but I'd rather not share the parameter generating code between these two API calls: They do very different things (even if the first few parameters are the same) and I'd find the above `Endpoint` enum to be a little harder to reason about when debugging. I might just be stuck in Objective-C brain though 😅 

> Though this wouldn't really work if the parameters could change between the initialization time and the function call time. 

Yeah, that's the main issue — the key could change between two subsequent requests, so I think it's clearer if `StripeAttest.keyId` is the source of truth instead of having the backend object store its own key. App ID and Device ID will never change during the lifetime of the process though.

Reply:
Makes sense! This definitely added some complexity that ultimately probably isn't worth it 👍